### PR TITLE
gpio: Add GPIO voltage overrides

### PR DESCRIPTION
**Pull request by brockus-zephyr**
Add GPIO_flag values for GPIO DTS creation that allows for pins that do not use the default voltage.  IT8xxx2 supports setting voltage levels for each pin. The default for this SoC is 3P3 but they can be set individually to 1P8.

This will be used, for example:
>         i2c_c_scl {
>                 gpios = <&gpioc 1 (GPIO_INPUT | GPIO_VOLTAGE_1P8)>;
>                 enum-name = "GPIO_I2C_C_SCL";
>                 label = "I2C_C_SCL";
>         };

### CODE REVIEW

**Review by keith-zephyr — APPROVED** [review]

**Review by sjg20 — APPROVED** [review]
I suggest _SHIFT is a better suffix for bitfield position than _POS

**Comment by keith-zephyr** [comment]
@carlescufi, can you review this PR?

**Comment by brockus-zephyr** [comment]
> Need @mnkp review as GPIO maintainer. Additionally, this isn't used anywhere so curious about its utility at this point.

The code that relies on this header change is waiting for this PR to land

**Comment by brockus-zephyr** [comment]
@galak any idea why buildkite/zephyr is failing?  This does not look related to my header file changes. @keith-zephyr kicked the build yesterday and that build did not appear to retry.  After ten minutes, it was showing that buildkite failed after over an hour.

**Comment by keith-zephyr** [comment]
@pabigot can you please review this PR.

**Review by pabigot — COMMENTED** [review]
I'm no longer a Zephyr maintainer; this needs input from @mnkp as the maintainer of the GPIO system, and possibly @anangl representing one of the SoCs where this isn't going to be a supported feature.

It also should be presented at an API meeting @carlescufi

File: include/dt-bindings/gpio/gpio.h
Comment (mnkp) [inline include/dt-bindings/gpio/gpio.h:81]: This flags are likely not to be supported on per pin basis by other SoC families. Let's be explicit and mention it in the comment here.
```
 * @name GPIO pin voltage flags
 * 
 * Only applicable if SoC allows to configure pin voltage per individual pin.
 * 
 * @{
```

**Review by mnkp — COMMENTED** [review]
The `include/dt-bindings/gpio/gpio.h` file is based on the Linux [gpio.h](https://github.com/torvalds/linux/blob/master/include/dt-bindings/gpio/gpio.h) which doesn't define flags to configure pin voltage level. In principle we want to keep our DTS close to the Linux one. The original idea was to have `include/dt-bindings/gpio/gpio.h` identical to its Linux counterpart.

That said, it doesn't look like Linux allows to configure pin voltage level in DTS at all. Seems like we are free to choose how we want to implement it.

The approach proposed in this PR looks reasonable. Similarly to pin drive strength or pull up, pull down configuration defined as flags in `include/dt-bindings/gpio/gpio.h` the voltage level describes physical quality of the pin. The only problem I see is that unlike drive strength or pin bias it is likely that the next SoC family will not allow to configure voltage per individual pin but rather per bunk / port. The flags will not be usable in that case. Let's clarify this in the doxygen documentation.

File: include/dt-bindings/gpio/gpio.h
Comment (brockus-zephyr) [reply include/dt-bindings/gpio/gpio.h:84]: Done

File: include/dt-bindings/gpio/gpio.h
Comment (brockus-zephyr) [reply include/dt-bindings/gpio/gpio.h]: Changes made. Thank you Piotr for the review.

**Review by keith-zephyr — APPROVED** [review]

**Comment by carlescufi** [comment]
@mnkp can you please revisit?

**Review by mnkp — APPROVED** [review]
The GPIO voltage overrides introduced by this PR are vendor specific, such things are normally handled in DTS by vendor specific properties. This is not possible in case of GPIO since we pass configuration parameters via `#gpio-cells` specifier and not DTS properties. The solution proposed in this PR is fine with me but it would be good to have additional approval by people working with DTS @galak @mbolivar-nordic @carlescufi Alternatively we could also discuss it shortly at the API meeting.

**Comment by mnkp** [comment]
@brockus-zephyr Let's add description to this PR. You can copy / paste the commit message. It will make it easier for anyone that comes across this PR in the future.

File: include/dt-bindings/gpio/gpio.h
Comment (anangl) [inline include/dt-bindings/gpio/gpio.h:105]: Why aren't those suffixes `1V8`, `3V3`, and `5V0`? Such notation seems to be more widely used in Zephyr.

Comment (brockus-zephyr) [reply include/dt-bindings/gpio/gpio.h:105]: I have seen EEs use 1P8, 3P3 and 5P0 when the name is already clear this is a voltage. On the other hand, if the names were GPIO_LEVEL_* then it would not be clear and 1V8, 3V3 and 5V0 would be used to make it clear what the level is referring to. I can change this if it is desired to move away from this EE naming.

**Comment by brockus-zephyr** [comment]
> @brockus-zephyr Let's add description to this PR. You can copy / paste the commit message. It will make it easier for anyone that comes across this PR in the future.

@mnkp Where would you like me to add the commit message? Something like the following?

/**
 * @name GPIO pin voltage flags
 *
 * Only applicable if SoC allows to configure pin voltage per individual pin.
 *
 * @{
 */

/*
 * History from commit message:
 *
 * Add GPIO_flag values for GPIO DTS creation that allows for pins that do not
 * use the default voltage.  IT8xxx2 supports setting voltage levels for each
 * pin. The default for this SoC is 3P3 but they can be set individually to
 * 1P8.
 *
 * This will be used, for example:
 *
 *	i2c_c_scl {
 *		gpios = <&gpioc 1 (GPIO_INPUT | GPIO_VOLTAGE_1P8)>;
 *		enum-name = "GPIO_I2C_C_SCL";
 *		label = "I2C_C_SCL";
 *	};
 */

/**
 * @cond INTERNAL_HIDDEN
 *
 * For internal use only, skip these in public documentation.
 */
#define GPIO_VOLTAGE_POS        6
#define GPIO_VOLTAGE_MASK       (3U << GPIO_VOLTAGE_POS)
/**
 * @endcond
 */

/** Set pin at the default voltage level */
#define GPIO_VOLTAGE_DEFAULT    (0U << GPIO_VOLTAGE_POS)
/** Set pin voltage level at 1.8 V */
#define GPIO_VOLTAGE_1P8        (1U << GPIO_VOLTAGE_POS)
/** Set pin voltage level at 3.3 V */
#define GPIO_VOLTAGE_3P3        (2U << GPIO_VOLTAGE_POS)
/** Set pin voltage level at 5.0 V */
#define GPIO_VOLTAGE_5P0        (3U << GPIO_VOLTAGE_POS)

/** @} */

File: include/dt-bindings/gpio/gpio.h
Comment (brockus-zephyr) [reply include/dt-bindings/gpio/gpio.h:105]: I grepped the code and there are uses of both P and V.

**Comment by galak** [comment]
@nashif I think this needed more discussion.

**Comment by brockus-zephyr** [comment]
Reverted in https://github.com/zephyrproject-rtos/zephyr/pull/36150
Resubmitted in https://github.com/zephyrproject-rtos/zephyr/pull/36179

**Pull request merged**
